pyro/distributions/transforms/spline.py

```diff
@@ -0,0 +1,352 @@
+    logabsdet = torch.zeros_like(inputs)
+
+    # For numerical stability, put lower/upper limits on parameters. E.g .give every bin min_bin_width,
+    # then add width fraction of remaining length
+    # NOTE: Do this here rather than higher up because we want everything to ensure numerical
+    # stability within this function
+    widths = min_bin_width + (1.0 - min_bin_width * num_bins) * widths
+    heights = min_bin_height + (1.0 - min_bin_height * num_bins) * heights
+    derivatives = min_derivative + derivatives
+    lambdas = (1 - 2 * min_lambda) * lambdas + min_lambda
+
+    # Cumulative widths are x (y for inverse) position of knots
+    # Similarly, cumulative heights are y (x for inverse) position of knots
+    widths, cumwidths = calculate_knots(widths, left, right)
+    heights, cumheights = calculate_knots(heights, bottom, top)
+
+    # Pad left and right derivatives with fixed values at first and last knots
+    # These are 1 since the function is the identity outside the bounding box and the derivative is continuous
+    # NOTE: Not sure why this is 1.0 - min_derivative rather than 1.0. I've copied this from original implementation
+    derivatives = F.pad(derivatives, pad=(1, 1), mode='constant', value=1.0 - min_derivative)
+
+    # Get the index of the bin that each input is in
+    # bin_idx ~ (batch_dim, input_dim, 1)
+    bin_idx = search_sorted(cumheights if inverse else cumwidths, inputs)[..., None]
+
+    # Select the value for the relevant bin for the variables used in the main calculation
+    input_widths = select_bins(widths, bin_idx)
+    input_cumwidths = select_bins(cumwidths, bin_idx)
+    input_cumheights = select_bins(cumheights, bin_idx)
+    input_delta = select_bins(heights / widths, bin_idx)
+    input_derivatives = select_bins(derivatives, bin_idx)
+    input_derivatives_plus_one = select_bins(derivatives[..., 1:], bin_idx)
+    input_heights = select_bins(heights, bin_idx)
+    input_lambdas = select_bins(lambdas, bin_idx)
+
+    # The weight, w_a, at the left-hand-side of each bin
+    # We are free to choose w_a, so set it to 1
+    wa = 1.0
+
+    # The weight, w_b, at the right-hand-side of each bin
+    # This turns out to be a multiple of the w_a
+    # TODO: Should this be done in log space for numerical stability?
+    wb = torch.sqrt(input_derivatives / input_derivatives_plus_one) * wa
+
+    # The weight, w_c, at the division point of each bin
+    # Recall that each bin is divided into two parts so we have enough d.o.f. to fit spline
+    wc = (input_lambdas * wa * input_derivatives + (1 - input_lambdas) * wb * input_derivatives_plus_one) / input_delta
+
+    # Calculate y coords of bins
+    ya = input_cumheights
+    yb = input_heights + input_cumheights
+    yc = ((1.0 - input_lambdas) * wa * ya + input_lambdas * wb * yb) / ((1.0 - input_lambdas) * wa + input_lambdas * wb)
+
+    # The core monotonic rational spline equation
+    if inverse:
+        # TODO: Comments!
```

Comment:
nit: no need to add this todo: uncommented code already screams out "todo comments!" in its own voice

Reply:
Removed